nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/StreamBulkScroller.java

```diff
@@ -135,50 +137,45 @@ protected void processRecord(ComputationContext context, Record record) {
                     long documentCount = 0;
                     while (scroll.hasResults()) {
                         List<String> docIds = scroll.getResults();
-                        // send these ids as keys to the appropriate stream
-                        // key will be bulkOperationId/docId
-                        // value/data is a BulkCommand serialized as JSON
-                        docIds.forEach(docId -> context.produceRecord(command.getOperation(), record.key + '/' + docId,
-                                record.data));
+                        documentIds.addAll(docIds);
+                        if (documentIds.size() >= recordBatchSize) {
+                            // send these ids as keys to the appropriate stream
+                            // key will be bulkActionId/docIds
+                            // value/data is a BulkCommand serialized as JSON
+                            produceRecord(context, command.getAction(), bulkActionId, record.getData());
+                        }
+
                         documentCount += docIds.size();
-                        context.askForCheckpoint();
                         // next batch
                         scroll = session.scroll(scroll.getScrollId());
                         TransactionHelper.commitOrRollbackTransaction();
                         TransactionHelper.startTransaction();
                     }
-                    kvStore.put(record.key + STATE, COMPLETED.toString());
-                    kvStore.put(record.key + SCROLLED_DOCUMENT_COUNT, documentCount);
+
+                    while (!documentIds.isEmpty()) {
```

Comment:
not sure to understand why you need to loop here

Reply:
We don't indeed.